File: src/daemon-client.ts
/**
 * MCP-CLI Daemon Client - IPC client for communicating with daemon workers
 * 
 * Handles spawning daemons, detecting stale connections, and forwarding requests.
 */

import { existsSync } from 'node:fs';
import { join } from 'node:path';
import {
  type ServerConfig,
  debug,
  getSocketPath,
  getConfigHash,
  getSocketDir,
} from './config.js';
import {
  readPidFile,
  removePidFile,
  removeSocketFile,
  isProcessRunning,
  killProcess,
  type DaemonRequest,
  type DaemonResponse,
} from './daemon.js';

// ============================================================================
// Daemon Connection
// ============================================================================

/**
 * Represents a daemon connection for a specific server
 */
export interface DaemonConnection {
  serverName: string;
  listTools: () => Promise<unknown>;
  callTool: (toolName: string, args: Record<string, unknown>) => Promise<unknown>;
  getInstructions: () => Promise<string | undefined>;
  close: () => Promise<void>;
}

/**
 * Generate a unique request ID
 */
function generateRequestId(): string {
  return `${Date.now()}-${Math.random().toString(36).slice(2, 9)}`;
}

/**
 * Send a request to the daemon and wait for response
 */
async function sendRequest(socketPath: string, request: DaemonRequest): Promise<DaemonResponse> {
  return new Promise((resolve, reject) => {
    const socket = Bun.connect({
      unix: socketPath,
      socket: {
        open(socket) {
          socket.write(JSON.stringify(request));
        },
        data(socket, data) {
          try {
            const response = JSON.parse(data.toString().trim());
            socket.end();
            resolve(response);
          } catch (err) {
            socket.end();
            reject(new Error('Invalid response from daemon'));
          }
        },
        error(socket, error) {
          reject(error);
        },
        close() {
          // Connection closed
        },
        connectError(socket, error) {
          reject(error);
        },
      },
    });

    // Timeout after 5 seconds (fast fallback to direct connection)
    setTimeout(() => {
      reject(new Error('Daemon request timeout'));
    }, 5000);
  });
}

/**
 * Check if daemon is running and has matching config
 */
function isDaemonValid(serverName: string, config: ServerConfig): boolean {
  const socketPath = getSocketPath(serverName);
  const pidInfo = readPidFile(serverName);

  // No PID file = no daemon
  if (!pidInfo) {
    debug(`[daemon-client] No PID file for ${serverName}`);
    return false;
  }

  // Check if process is actually running
  if (!isProcessRunning(pidInfo.pid)) {
    debug(`[daemon-client] Process ${pidInfo.pid} not running, cleaning up`);
    removePidFile(serverName);
    removeSocketFile(serverName);
    return false;
  }

  // Check if config matches
  const currentHash = getConfigHash(config);
  if (pidInfo.configHash !== currentHash) {
    debug(`[daemon-client] Config hash mismatch for ${serverName}, killing old daemon`);
    killProcess(pidInfo.pid);
    removePidFile(serverName);
    removeSocketFile(serverName);
    return false;
  }

  // Check if socket exists
  if (!existsSync(socketPath)) {
    debug(`[daemon-client] Socket missing for ${serverName}, cleaning up`);
    killProcess(pidInfo.pid);
    removePidFile(serverName);
    return false;
  }

  return true;
}

/**
 * Spawn a new daemon process for a server
 */
async function spawnDaemon(serverName: string, config: ServerConfig): Promise<boolean> {
  debug(`[daemon-client] Spawning daemon for ${serverName}`);

  // Find the daemon script path
  const daemonScript = join(import.meta.dir, 'daemon.ts');

  const configJson = JSON.stringify(config);

  // Spawn detached process
  const proc = Bun.spawn({
    cmd: ['bun', 'run', daemonScript, '--daemon', serverName, configJson],
    stdout: 'pipe',
    stderr: 'pipe',
    env: { ...process.env },
  });

  // Wait for daemon to signal readiness or fail
  return new Promise((resolve) => {
    let resolved = false;

    const reader = proc.stdout.getReader();

    const checkReady = async () => {
      try {
        const { value, done } = await reader.read();
        if (done) {
          if (!resolved) {
            resolved = true;
            resolve(false);
          }
          return;
        }

        const text = new TextDecoder().decode(value);
        if (text.includes('DAEMON_READY')) {
          if (!resolved) {
            resolved = true;
            // Don't await the process, let it run detached
            proc.unref();
            resolve(true);
          }
        } else {
          // Keep reading
          checkReady();
        }
      } catch {
        if (!resolved) {
          resolved = true;
          resolve(false);
        }
      }
    };

    checkReady();

    // Timeout after 5 seconds (fast fallback to direct connection)
    setTimeout(() => {
      if (!resolved) {
        resolved = true;
        debug(`[daemon-client] Daemon spawn timeout for ${serverName}`);
        resolve(false);
      }
    }, 5000);

    // Check for early exit
    proc.exited.then((code) => {
      if (!resolved && code !== 0) {
        resolved = true;
        debug(`[daemon-client] Daemon exited with code ${code}`);
        resolve(false);
      }
    });
  });
}

/**
 * Get or create a daemon connection for a server
 * Returns null if daemon mode fails (caller should fallback to direct connection)
 */
export async function getDaemonConnection(
  serverName: string,
  config: ServerConfig,
): Promise<DaemonConnection | null> {
  const socketPath = getSocketPath(serverName);

  // Check if valid daemon exists
  if (!isDaemonValid(serverName, config)) {
    // Spawn new daemon
    const spawned = await spawnDaemon(serverName, config);
    if (!spawned) {
      debug(`[daemon-client] Failed to spawn daemon for ${serverName}`);
      return null;
    }

    // Wait a bit for socket to be ready
    await new Promise((r) => setTimeout(r, 100));
  }

  // Verify socket exists
  if (!existsSync(socketPath)) {
    debug(`[daemon-client] Socket not found after spawn for ${serverName}`);
    return null;
  }

  // Test connection with ping
  try {
    const pingResponse = await sendRequest(socketPath, {
      id: generateRequestId(),
      type: 'ping',
    });

    if (!pingResponse.success) {
      debug(`[daemon-client] Ping failed for ${serverName}`);
      return null;
    }
  } catch (error) {
    debug(`[daemon-client] Connection test failed for ${serverName}: ${(error as Error).message}`);
    return null;
  }

  debug(`[daemon-client] Connected to daemon for ${serverName}`);

  // Return connection interface
  return {
    serverName,

    async listTools(): Promise<unknown> {
      const response = await sendRequest(socketPath, {
        id: generateRequestId(),
        type: 'listTools',
      });

      if (!response.success) {
        throw new Error(response.error?.message ?? 'listTools failed');
      }

      return response.data;
    },

    async callTool(toolName: string, args: Record<string, unknown>): Promise<unknown> {
      const response = await sendRequest(socketPath, {
        id: generateRequestId(),
        type: 'callTool',
        toolName,
        args,
      });

      if (!response.success) {
        throw new Error(response.error?.message ?? 'callTool failed');
      }

      return response.data;
    },

    async getInstructions(): Promise<string | undefined> {
      const response = await sendRequest(socketPath, {
        id: generateRequestId(),
        type: 'getInstructions',
      });

      if (!response.success) {
        throw new Error(response.error?.message ?? 'getInstructions failed');
      }

      return response.data as string | undefined;
    },

    async close(): Promise<void> {
      // Just disconnect, don't tell daemon to close (let it idle timeout)
      debug(`[daemon-client] Disconnecting from ${serverName} daemon`);
    },
  };
}

/**
 * Clean up any orphaned daemon processes and sockets
 * Call this on CLI startup
 */
export async function cleanupOrphanedDaemons(): Promise<void> {
  const socketDir = getSocketDir();

  if (!existsSync(socketDir)) {
    return;
  }

  try {
    const files = await Array.fromAsync(new Bun.Glob('*.pid').scan(socketDir));

    for (const file of files) {
      const serverName = file.replace('.pid', '');
      const pidInfo = readPidFile(serverName);

      if (pidInfo && !isProcessRunning(pidInfo.pid)) {
        debug(`[daemon-client] Cleaning up orphaned daemon: ${serverName}`);
        removePidFile(serverName);
        removeSocketFile(serverName);
      }
    }
  } catch {
    // Ignore errors during cleanup scan
  }
}

File: src/daemon_client.rs
use crate::config::{
    debug, get_config_hash, get_socket_dir, get_socket_path, ServerConfig,
};
use crate::daemon::{
    is_process_running, kill_process, read_pid_file, remove_pid_file, remove_socket_file,
    DaemonRequest, DaemonResponse,
};
use crate::errors::CliError;
use std::fs;
use std::path::PathBuf;
use std::time::Duration;
use tokio::io::{AsyncBufReadExt, AsyncReadExt, AsyncWriteExt};
use tokio::net::UnixStream;

// ============================================================================
// Daemon Connection
// ============================================================================

pub struct DaemonConnection {
    pub server_name: String,
    pub socket_path: PathBuf,
}

fn generate_request_id() -> String {
    let now = chrono::Utc::now().timestamp_millis();
    let rand_val: u32 = rand::random();
    format!("{}-{}", now, rand_val)
}

async fn send_request(
    socket_path: &std::path::Path,
    request: DaemonRequest,
) -> Result<DaemonResponse, CliError> {
    let connect_fut = UnixStream::connect(socket_path);

    let mut stream = tokio::select! {
        res = connect_fut => {
            res.map_err(|e| crate::errors::CliError {
                code: crate::errors::ErrorCode::NetworkError,
                error_type: "DAEMON_CONNECT_FAILED".to_string(),
                message: format!("Failed to connect to daemon socket: {}", e),
                details: None,
                suggestion: None,
            })?
        }
        _ = tokio::time::sleep(Duration::from_secs(5)) => {
            return Err(crate::errors::CliError {
                code: crate::errors::ErrorCode::NetworkError,
                error_type: "DAEMON_TIMEOUT".to_string(),
                message: "Daemon connection timeout".to_string(),
                details: None,
                suggestion: None,
            });
        }
    };

    let serialized = serde_json::to_vec(&request).map_err(|e| crate::errors::CliError {
        code: crate::errors::ErrorCode::ClientError,
        error_type: "DAEMON_SERIALIZE_FAILED".to_string(),
        message: format!("Failed to serialize request: {}", e),
        details: None,
        suggestion: None,
    })?;

    if let Err(e) = stream.write_all(&serialized).await {
        return Err(crate::errors::CliError {
            code: crate::errors::ErrorCode::NetworkError,
            error_type: "DAEMON_WRITE_FAILED".to_string(),
            message: format!("Failed to write to daemon: {}", e),
            details: None,
            suggestion: None,
        });
    }
    if let Err(e) = stream.flush().await {
        return Err(crate::errors::CliError {
            code: crate::errors::ErrorCode::NetworkError,
            error_type: "DAEMON_WRITE_FAILED".to_string(),
            message: format!("Failed to flush daemon connection: {}", e),
            details: None,
            suggestion: None,
        });
    }

    let mut buffer = Vec::new();
    let read_fut = stream.read_to_end(&mut buffer);

    tokio::select! {
        res = read_fut => {
            res.map_err(|e| crate::errors::CliError {
                code: crate::errors::ErrorCode::NetworkError,
                error_type: "DAEMON_READ_FAILED".to_string(),
                message: format!("Failed to read from daemon: {}", e),
                details: None,
                suggestion: None,
            })?;
        }
        _ = tokio::time::sleep(Duration::from_secs(5)) => {
            return Err(crate::errors::CliError {
                code: crate::errors::ErrorCode::NetworkError,
                error_type: "DAEMON_TIMEOUT".to_string(),
                message: "Daemon response timeout".to_string(),
                details: None,
                suggestion: None,
            });
        }
    }

    let response: DaemonResponse = serde_json::from_slice(&buffer).map_err(|e| crate::errors::CliError {
        code: crate::errors::ErrorCode::ServerError,
        error_type: "DAEMON_RESPONSE_INVALID".to_string(),
        message: format!("Daemon returned invalid JSON: {}", e),
        details: Some(String::from_utf8_lossy(&buffer).to_string()),
        suggestion: None,
    })?;

    Ok(response)
}

fn is_daemon_valid(server_name: &str, config: &ServerConfig) -> bool {
    let socket_path = get_socket_path(server_name);
    let pid_info = read_pid_file(server_name);

    if pid_info.is_none() {
        debug(&format!("[daemon-client] No PID file for {}", server_name));
        return false;
    }
    let pid_info = pid_info.unwrap();

    if !is_process_running(pid_info.pid) {
        debug(&format!(
            "[daemon-client] Process {} not running, cleaning up",
            pid_info.pid
        ));
        remove_pid_file(server_name);
        remove_socket_file(server_name);
        return false;
    }

    let current_hash = get_config_hash(config);
    if pid_info.config_hash != current_hash {
        debug(&format!(
            "[daemon-client] Config hash mismatch for {}, killing old daemon",
            server_name
        ));
        kill_process(pid_info.pid);
        remove_pid_file(server_name);
        remove_socket_file(server_name);
        return false;
    }

    if !socket_path.exists() {
        debug(&format!(
            "[daemon-client] Socket missing for {}, cleaning up",
            server_name
        ));
        kill_process(pid_info.pid);
        remove_pid_file(server_name);
        return false;
    }

    true
}

async fn spawn_daemon(server_name: &str, config: &ServerConfig) -> bool {
    debug(&format!("[daemon-client] Spawning daemon for {}", server_name));

    let current_exe = match std::env::current_exe() {
        Ok(path) => path,
        Err(e) => {
            debug(&format!("[daemon-client] Failed to get current_exe: {}", e));
            return false;
        }
    };

    let config_json = match serde_json::to_string(config) {
        Ok(json) => json,
        Err(e) => {
            debug(&format!("[daemon-client] Failed to serialize config: {}", e));
            return false;
        }
    };

    // Spawn detached process
    let mut child = match tokio::process::Command::new(current_exe)
        .args(&["--daemon", server_name, &config_json])
        .stdin(std::process::Stdio::null())
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::null()) // Don't forward stderr to console for daemon
        .spawn()
    {
        Ok(c) => c,
        Err(e) => {
            debug(&format!(
                "[daemon-client] Failed to spawn child process: {}",
                e
            ));
            return false;
        }
    };

    let stdout = child.stdout.take().unwrap();
    let mut reader = tokio::io::BufReader::new(stdout);
    let mut line = String::new();

    let read_ready = async {
        loop {
            line.clear();
            match reader.read_line(&mut line).await {
                Ok(0) => return false, // EOF
                Ok(_) => {
                    if line.contains("DAEMON_READY") {
                        return true;
                    }
                }
                Err(_) => return false,
            }
        }
    };

    let spawned = tokio::select! {
        ready = read_ready => ready,
        _ = tokio::time::sleep(Duration::from_secs(5)) => {
            debug(&format!("[daemon-client] Daemon spawn timeout for {}", server_name));
            false
        }
    };

    if spawned {
        debug(&format!(
            "[daemon-client] Daemon successfully spawned for {}",
            server_name
        ));
        true
    } else {
        let _ = child.kill().await;
        false
    }
}

pub async fn get_daemon_connection(
    server_name: &str,
    config: &ServerConfig,
) -> Option<DaemonConnection> {
    let socket_path = get_socket_path(server_name);

    if !is_daemon_valid(server_name, config) {
        let spawned = spawn_daemon(server_name, config).await;
        if !spawned {
            debug(&format!(
                "[daemon-client] Failed to spawn daemon for {}",
                server_name
            ));
            return None;
        }
        tokio::time::sleep(Duration::from_millis(100)).await;
    }

    if !socket_path.exists() {
        debug(&format!(
            "[daemon-client] Socket not found after spawn for {}",
            server_name
        ));
        return None;
    }

    // Ping test
    match send_request(
        &socket_path,
        DaemonRequest {
            id: generate_request_id(),
            req_type: "ping".to_string(),
            tool_name: None,
            args: None,
        },
    )
    .await
    {
        Ok(resp) => {
            if !resp.success {
                debug(&format!("[daemon-client] Ping failed for {}", server_name));
                return None;
            }
        }
        Err(e) => {
            debug(&format!(
                "[daemon-client] Connection test failed for {}: {}",
                server_name, e
            ));
            return None;
        }
    }

    debug(&format!(
        "[daemon-client] Connected to daemon for {}",
        server_name
    ));
    Some(DaemonConnection {
        server_name: server_name.to_string(),
        socket_path,
    })
}

pub async fn cleanup_orphaned_daemons() {
    let socket_dir = get_socket_dir();
    if !socket_dir.exists() {
        return;
    }

    if let Ok(entries) = fs::read_dir(socket_dir) {
        for entry in entries.filter_map(Result::ok) {
            let path = entry.path();
            if path.extension().and_then(|s| s.to_str()) == Some("pid") {
                if let Some(file_name) = path.file_stem().and_then(|s| s.to_str()) {
                    let server_name = file_name;
                    if let Some(pid_info) = read_pid_file(server_name) {
                        if !is_process_running(pid_info.pid) {
                            debug(&format!(
                                "[daemon-client] Cleaning up orphaned daemon: {}",
                                server_name
                            ));
                            remove_pid_file(server_name);
                            remove_socket_file(server_name);
                        }
                    }
                }
            }
        }
    }
}

impl DaemonConnection {
    pub async fn close(self) -> Result<(), CliError> {
        debug(&format!(
            "[daemon-client] Disconnecting from {} daemon",
            self.server_name
        ));
        Ok(())
    }

    pub async fn list_tools(&self) -> Result<serde_json::Value, CliError> {
        let response = send_request(
            &self.socket_path,
            DaemonRequest {
                id: generate_request_id(),
                req_type: "listTools".to_string(),
                tool_name: None,
                args: None,
            },
        )
        .await?;

        if !response.success {
            return Err(crate::errors::CliError {
                code: crate::errors::ErrorCode::ServerError,
                error_type: "DAEMON_ERROR".to_string(),
                message: response
                    .error
                    .map(|e| e.message)
                    .unwrap_or_else(|| "listTools failed".to_string()),
                details: None,
                suggestion: None,
            });
        }

        response.data.ok_or_else(|| crate::errors::CliError {
            code: crate::errors::ErrorCode::ServerError,
            error_type: "DAEMON_ERROR".to_string(),
            message: "Daemon listTools returned no data".to_string(),
            details: None,
            suggestion: None,
        })
    }

    pub async fn call_tool(
        &self,
        tool_name: &str,
        args: serde_json::Value,
    ) -> Result<serde_json::Value, CliError> {
        let response = send_request(
            &self.socket_path,
            DaemonRequest {
                id: generate_request_id(),
                req_type: "callTool".to_string(),
                tool_name: Some(tool_name.to_string()),
                args: Some(args),
            },
        )
        .await?;

        if !response.success {
            return Err(crate::errors::CliError {
                code: crate::errors::ErrorCode::ServerError,
                error_type: "DAEMON_ERROR".to_string(),
                message: response
                    .error
                    .map(|e| e.message)
                    .unwrap_or_else(|| "callTool failed".to_string()),
                details: None,
                suggestion: None,
            });
        }

        response.data.ok_or_else(|| crate::errors::CliError {
            code: crate::errors::ErrorCode::ServerError,
            error_type: "DAEMON_ERROR".to_string(),
            message: "Daemon call_tool returned no data".to_string(),
            details: None,
            suggestion: None,
        })
    }

    pub async fn get_instructions(&self) -> Result<Option<String>, CliError> {
        let response = send_request(
            &self.socket_path,
            DaemonRequest {
                id: generate_request_id(),
                req_type: "getInstructions".to_string(),
                tool_name: None,
                args: None,
            },
        )
        .await?;

        if !response.success {
            return Err(crate::errors::CliError {
                code: crate::errors::ErrorCode::ServerError,
                error_type: "DAEMON_ERROR".to_string(),
                message: response
                    .error
                    .map(|e| e.message)
                    .unwrap_or_else(|| "getInstructions failed".to_string()),
                details: None,
                suggestion: None,
            });
        }

        if let Some(data) = response.data {
            if data.is_null() {
                Ok(None)
            } else if let Some(s) = data.as_str() {
                Ok(Some(s.to_string()))
            } else {
                Ok(Some(data.to_string()))
            }
        } else {
            Ok(None)
        }
    }
}
